File: src/discord/webhook.ts
import { logger } from "../utils/logger";
import type { DiscordEmbed } from "./embed";

/**
 * @description Discord Webhookのペイロード
 * @property embeds - 送信するEmbedの配列
 * @property username - Webhook表示名 @optional
 * @property avatar_url - Webhookアバター画像URL @optional
 */
export interface WebhookPayload {
  embeds: DiscordEmbed[];
  username?: string;
  avatar_url?: string;
}

/**
 * @description 配信者情報(Webhook表示用)
 * @property displayName - 表示名
 * @property profileImageUrl - プロフィール画像URL
 */
export interface StreamerInfo {
  displayName: string;
  profileImageUrl: string;
}

/**
 * @description 単一のWebhookにEmbedを送信
 * @param webhookUrl - Discord Webhook URL
 * @param embed - 送信するEmbed
 * @param streamer - 配信者情報
 * @throws 送信失敗時にエラー
 */
export async function sendWebhook(
  webhookUrl: string,
  embed: DiscordEmbed,
  streamer: StreamerInfo
): Promise<void> {
  const payload: WebhookPayload = {
    embeds: [embed],
    username: streamer.displayName,
    avatar_url: streamer.profileImageUrl,
  };

  const response = await fetch(webhookUrl, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });

  if (!response.ok) {
    const error = await response.text();
    throw new Error(`Webhook送信失敗: ${response.status} ${error}`);
  }

  logger.debug(`Webhook送信成功: ${webhookUrl.slice(0, 50)}...`);
}

/**
 * @description 複数のWebhookにEmbedを並列送信
 * @param webhookUrls - Discord Webhook URLの配列
 * @param embed - 送信するEmbed
 * @param streamer - 配信者情報
 */
export async function sendToMultipleWebhooks(
  webhookUrls: string[],
  embed: DiscordEmbed,
  streamer: StreamerInfo
): Promise<void> {
  const results = await Promise.allSettled(
    webhookUrls.map((url) => sendWebhook(url, embed, streamer))
  );

  for (const [index, result] of results.entries()) {
    if (result.status === "rejected") {
      logger.error(`Webhook送信エラー [${index + 1}/${webhookUrls.length}]:`, result.reason);
    }
  }
}

File: internal/discord/webhook.go
package discord

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"sync"
	"time"
)

// WebhookPayload はDiscord Webhookのペイロード。
type WebhookPayload struct {
	Embeds    []Embed `json:"embeds"`
	Username  string  `json:"username,omitempty"`
	AvatarURL string  `json:"avatar_url,omitempty"`
}

// StreamerInfo は配信者情報(Webhook表示用)。
type StreamerInfo struct {
	DisplayName     string
	ProfileImageURL string
}

// SendWebhook は単一のWebhookにEmbedを送信する。
func SendWebhook(ctx context.Context, webhookURL string, embed Embed, streamer StreamerInfo) error {
	payload := WebhookPayload{
		Embeds:    []Embed{embed},
		Username:  streamer.DisplayName,
		AvatarURL: streamer.ProfileImageURL,
	}

	body, err := json.Marshal(payload)
	if err != nil {
		return fmt.Errorf("WebhookペイロードのJSON変換に失敗: %w", err)
	}

	ctx, cancel := context.WithTimeout(ctx, 30*time.Second)
	defer cancel()

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, webhookURL, bytes.NewReader(body))
	if err != nil {
		return fmt.Errorf("Webhookリクエスト作成に失敗: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return fmt.Errorf("Webhook送信に失敗: %w", err)
	}
	defer resp.Body.Close()

	// レスポンスボディを消費してリソースを解放
	respBody, _ := io.ReadAll(resp.Body)

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return fmt.Errorf("Webhook送信失敗: %d %s", resp.StatusCode, string(respBody))
	}

	slog.Debug("Webhook送信成功", "url", truncate(webhookURL, 50))
	return nil
}

// SendToMultipleWebhooks は複数のWebhookにEmbedを並列送信する。
func SendToMultipleWebhooks(ctx context.Context, webhookURLs []string, embed Embed, streamer StreamerInfo) {
	var wg sync.WaitGroup
	for i, url := range webhookURLs {
		wg.Add(1)
		go func(idx int, u string) {
			defer wg.Done()
			if err := SendWebhook(ctx, u, embed, streamer); err != nil {
				slog.Error("Webhook送信エラー",
					"index", idx+1,
					"total", len(webhookURLs),
					"error", err)
			}
		}(i, url)
	}
	wg.Wait()
}

// truncate は文字列を指定長で切り詰める。
func truncate(s string, maxLen int) string {
	if len(s) <= maxLen {
		return s
	}
	return s[:maxLen] + "..."
}
